# Jumpman23 Analysis

In [1]:
import pandas as pd
import numpy as np

dataset = pd.read_csv('analyze_me.csv',parse_dates = ['when_the_delivery_started','when_the_Jumpman_arrived_at_pickup',
                                                  'when_the_Jumpman_left_pickup','when_the_Jumpman_arrived_at_dropoff'])

In [2]:
import datetime
import googlemaps
import gmplot
import plotly.offline as py 
import plotly.graph_objs as go
import cufflinks as cf
import plotly.figure_factory as ff
cf.go_offline()
py.init_notebook_mode(connected=True)

In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5983 entries, 0 to 5982
Data columns (total 18 columns):
delivery_id                            5983 non-null int64
customer_id                            5983 non-null int64
jumpman_id                             5983 non-null int64
vehicle_type                           5983 non-null object
pickup_place                           5983 non-null object
place_category                         5100 non-null object
item_name                              4753 non-null object
item_quantity                          4753 non-null float64
item_category_name                     4753 non-null object
how_long_it_took_to_order              3038 non-null object
pickup_lat                             5983 non-null float64
pickup_lon                             5983 non-null float64
dropoff_lat                            5983 non-null float64
dropoff_lon                            5983 non-null float64
when_the_delivery_started              5983 non-nul

In [4]:
dataset.head()

delivery_id  customer_id  jumpman_id vehicle_type         pickup_place  \
0      1457973       327168      162381          van            Melt Shop   
1      1377056        64452      104533      bicycle  Prince Street Pizza   
2      1476547        83095      132725      bicycle           Bareburger   
3      1485494       271149      157175      bicycle          Juice Press   
4      1327707       122609      118095      bicycle    Blue Ribbon Sushi   

  place_category                    item_name  item_quantity  \
0       American                     Lemonade            1.0   
1          Pizza        Neapolitan Rice Balls            3.0   
2         Burger                   Bare Sodas            1.0   
3      Juice Bar      OMG! My Favorite Juice!            1.0   
4       Japanese  Spicy Tuna & Tempura Flakes            2.0   

     item_category_name how_long_it_took_to_order  pickup_lat  pickup_lon  \
0             Beverages           00:19:58.582052   40.744607  -73.990742   
1            Munchables           00:25:09.107093   40.723080  -73.994615   
2                Drinks           00:06:44.541717   40.728478  -73.998392   
3   Cold Pressed Juices                       NaN   40.738868  -74.002747   
4  Maki (Special Rolls)           00:03:45.035418   40.726110  -74.002492   

   dropoff_lat  dropoff_lon  when_the_delivery_started  \
0    40.752073   -73.985370 2014-10-26 13:51:59.898924   
1    40.719722   -73.991858 2014-10-16 21:58:58.654910   
2    40.728606   -73.995143 2014-10-28 21:39:52.654394   
3    40.751257   -74.005634 2014-10-30 10:54:11.531894   
4    40.709323   -74.015867 2014-10-10 00:07:18.450505   

  when_the_Jumpman_arrived_at_pickup when_the_Jumpman_left_pickup  \
0                                NaT                          NaT   
1         2014-10-16 22:26:02.120931   2014-10-16 22:48:23.091253   
2         2014-10-28 21:37:18.793405   2014-10-28 21:59:09.984810   
3         2014-10-30 11:04:17.759577   2014-10-30 11:16:37.895816   
4         2014-10-10 00:14:42.702223   2014-10-10 00:25:19.400294   

  when_the_Jumpman_arrived_at_dropoff  
0          2014-10-26 14:52:06.313088  
1          2014-10-16 22:59:22.948873  
2          2014-10-28 22:04:40.634962  
3          2014-10-30 11:32:38.090061  
4          2014-10-10 00:48:27.150595

* Delivery ID is duplicated on multiple rows, and for the duplicate rows, other than item and item type, all other columns remain the same
* This means that one customer is ordering different items from the same place in one order
### Creating new data frame 'df' to remove duplicated delivery IDs so that count of delivery ID signifies unique delivery/trips a Jumpan has to make

In [5]:
deliveryIDGroup = dataset.groupby('delivery_id')

In [6]:
df = deliveryIDGroup.max()
df.item_quantity = deliveryIDGroup.item_quantity.sum()
df.item_quantity = df.item_quantity.replace(0,np.nan)
df.reset_index(inplace=True)
df.head()

delivery_id  customer_id  jumpman_id vehicle_type           pickup_place  \
0      1271706        66556      154943          car           CVS Pharmacy   
1      1271751       104131      124897      bicycle            Duane Reade   
2      1271867       121706      124897      bicycle             McDonald's   
3      1272279       304449      112830      bicycle     Friend of a Farmer   
4      1272303       107621      127812      bicycle  Duane Reade duplicate   

      place_category           item_name  item_quantity item_category_name  \
0                NaN                 NaN            NaN                NaN   
1         Drug Store                 NaN            NaN                NaN   
2          Fast Food  World Famous Fries            2.0     Snacks & Sides   
3           American               Bacon            1.0              Sides   
4  Convenience Store                 NaN            NaN                NaN   

  how_long_it_took_to_order  pickup_lat  pickup_lon  dropoff_lat  dropoff_lon  \
0                       NaN   40.744157  -73.991987    40.744079   -73.991519   
1                       NaN   40.763217  -73.967613    40.765340   -73.967106   
2                       NaN   40.756799  -73.987899    40.756804   -73.998153   
3            00:09:08.75807   40.736794  -73.986800    40.733020   -73.982436   
4                       NaN   40.743171  -73.984004    40.744964   -73.985617   

   when_the_delivery_started when_the_Jumpman_arrived_at_pickup  \
0 2014-10-01 00:07:58.632482                                NaT   
1 2014-10-01 00:26:31.924774         2014-10-01 00:39:31.086322   
2 2014-10-01 01:00:06.756350         2014-10-01 01:19:29.205722   
3 2014-10-01 08:46:15.935061         2014-10-01 09:02:40.003541   
4 2014-10-01 09:20:21.573801         2014-10-01 09:26:01.194532   

  when_the_Jumpman_left_pickup when_the_Jumpman_arrived_at_dropoff  
0                          NaT          2014-10-01 00:30:21.109149  
1   2014-10-01 00:59:57.522402          2014-10-01 01:04:14.355157  
2   2014-10-01 01:36:49.131316          2014-10-01 01:49:29.034932  
3   2014-10-01 09:15:59.607582          2014-10-01 09:28:40.095456  
4   2014-10-01 09:37:56.158669          2014-10-01 09:39:41.631246

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5214 entries, 0 to 5213
Data columns (total 18 columns):
delivery_id                            5214 non-null int64
customer_id                            5214 non-null int64
jumpman_id                             5214 non-null int64
vehicle_type                           5214 non-null object
pickup_place                           5214 non-null object
place_category                         4373 non-null object
item_name                              3984 non-null object
item_quantity                          3984 non-null float64
item_category_name                     3984 non-null object
how_long_it_took_to_order              2579 non-null object
pickup_lat                             5214 non-null float64
pickup_lon                             5214 non-null float64
dropoff_lat                            5214 non-null float64
dropoff_lon                            5214 non-null float64
when_the_delivery_started              5214 non-nul

* 5214 unique orders

## Ordering Trends

In [8]:
def getDay(dateTime):
    return dateTime.strftime("%A")
def getDateFormat(dateTime):
    return dateTime.strftime("%d %b, %a")

In [9]:
df['delivery_start_date'] =df.when_the_delivery_started.apply(getDateFormat)

In [10]:
df['delivery_start_day']=df.when_the_delivery_started.apply(getDay)

In [11]:
df.delivery_start_date.value_counts().sort_index().iplot(mode='lines+markers', yTitle='Order Count', title = 'October 2018 Order Stats')

In [12]:
df.delivery_start_day.value_counts(normalize=1)

Sunday       0.181626
Thursday     0.151899
Wednesday    0.150173
Saturday     0.141542
Friday       0.136555
Tuesday      0.121596
Monday       0.116609
Name: delivery_start_day, dtype: float64

In [13]:
df[df.delivery_start_day=='Sunday'].delivery_start_date.value_counts().sort_index().iplot( kind = 'bar',yTitle='Order Count', title = 'Sunday Ordering Trend')

* Sunday is the most ordered day and we notice postive WoW growth rate for Sundays of week 3 and 4

### Hourly Demand

In [14]:
def getHours(dateTime):
    return dateTime.hour
df.when_the_delivery_started.apply(getHours).value_counts(normalize=1)

19    0.152474
18    0.123514
20    0.120061
21    0.075949
17    0.072114
12    0.070387
11    0.055044
13    0.054085
16    0.048140
14    0.045646
15    0.044304
22    0.036824
10    0.028577
23    0.022823
9     0.016110
0     0.012275
1     0.005946
8     0.005178
2     0.004219
7     0.003644
3     0.001343
4     0.000959
6     0.000384
Name: when_the_delivery_started, dtype: float64

In [15]:
df.when_the_delivery_started.apply(getHours).value_counts().iplot(kind='bar', title='Hourly Orders Breakdown')

* Time slot of  5 - 9pm has maximum demand with around 55% of total orders placed 

## Task - Time Analysis
   #### Time Taken to Place Order

In [16]:
df['how_long_it_took_to_order'] = pd.to_datetime(df.how_long_it_took_to_order, format = '%H:%M:%S.%f')

In [17]:
def minutes(dataset):
    return dataset.hour*60.0 + dataset.minute + dataset.second/60
df['min_to_order'] = df.how_long_it_took_to_order.apply(minutes)

In [18]:
df.min_to_order.describe()

count    2579.000000
mean        7.690507
std         5.712062
min         1.366667
25%         4.350000
50%         6.150000
75%         9.025000
max        73.216667
Name: min_to_order, dtype: float64

In [19]:
df.min_to_order.iplot(kind='box', title= 'Range of order times', yTitle = 'Time in minutes', xTitle = 'Minutes Taken to Place Order')

* Average customer ordering time is: 7.75 mins 
* Standard deviation in ordering is high at 5.31 mins
* Improving app design to make ordering easy can help cut down time taken by a user to complete an order


## Time taken to 1) reach pickup spot 2) wait at pickup 3) reach dropoff location

In [20]:
def total_minutes(dataset):
    return (dataset.total_seconds()/60)

In [21]:
start_to_pickup = df.when_the_Jumpman_arrived_at_pickup - df.when_the_delivery_started
pickup_duration = df.when_the_Jumpman_left_pickup - df.when_the_Jumpman_arrived_at_pickup
pickup_to_dropoff = df.when_the_Jumpman_arrived_at_dropoff-df.when_the_Jumpman_left_pickup

df['start_to_pickup_mins'] = start_to_pickup.apply(total_minutes)
df['pickup_duration_mins'] = pickup_duration.apply(total_minutes)
df['pickup_to_dropoff_mins'] = pickup_to_dropoff.apply(total_minutes)

In [22]:
df[['start_to_pickup_mins','pickup_duration_mins','pickup_to_dropoff_mins']].describe()

start_to_pickup_mins  pickup_duration_mins  pickup_to_dropoff_mins
count           4719.000000           4719.000000             4719.000000
mean              12.588740             18.294248               14.041480
std               12.231923             11.715884                9.314023
min               -2.651465              0.001877                0.839419
25%                3.509269             10.369364                7.938390
50%                9.876013             15.333771               11.658124
75%               18.749896             22.977090               17.304775
max              142.090125            267.954044              119.190060

In [23]:
df[['start_to_pickup_mins','pickup_duration_mins','pickup_to_dropoff_mins']].iplot(kind='box', title= 'Task-Time Analysis', yTitle = 'Time in minutes')

* High variability in picking up duration, long tail - Jumpmen are at pick up place for extended period of time. Huge potential for improvement
* Negative value of 'start to pickup time' indicates data integrity issue 

## Pickup Duration Analysis by Vendor 

#### Top 15 Vendors by Order Count

In [31]:
restaurants = df.groupby('pickup_place')
topRestaurants = restaurants.count().sort_values('delivery_id', ascending = 0).delivery_id[:15]
topRestaurants.iplot(kind='bar',title='Top Restaurants by Order Count')

* Top 15 restaurants account for 30% of all orders
* Pickup times at top restaurants need to be analyzed for potential improvemenet oportunities 

#### Pickup Time Analysis for Top 15 Vendors

In [48]:
a = restaurants.pickup_duration_mins.describe()

In [51]:
topVendorPickupTime = pd.DataFrame(columns = a.columns)
for i in range(len(topRestaurants)):
    topVendorPickupTime=topVendorPickupTime.append(a.loc[topRestaurants.index[i]])
topVendorPickupTime

count       mean        std        min  \
Shake Shack                            247.0  23.852485   9.806992   0.001877   
Momofuku Milk Bar                      156.0  15.617374   8.725106   1.843146   
The Meatball Shop                      147.0  13.853508   7.782065   1.334578   
sweetgreen                             125.0  19.389274  12.063139   5.318882   
Blue Ribbon Fried Chicken              110.0  18.891400  24.608881   6.717262   
Blue Ribbon Sushi                       93.0  20.358701  10.990109   4.827388   
Parm                                    84.0  17.105250   9.251719   4.852603   
Whole Foods Market                      73.0  37.077934  19.908625   8.678141   
Chipotle Mexican Grill $1.99 Delivery   63.0  14.501747   5.320113   7.329885   
Mighty Quinn's BBQ                      68.0  19.621718   7.665313   1.831372   
RedFarm Broadway                        68.0  20.777992  11.754326   4.782190   
Momofuku Noodle Bar                     61.0  31.500445  10.821636  11.324913   
Duane Reade                             44.0  13.663386   7.269580   0.821377   
McDonald's                              38.0  13.385864   4.510059   0.329697   
TAO                                     51.0  25.603544  15.126654   7.337831   

                                             25%        50%        75%  \
Shake Shack                            17.331913  22.829802  28.818984   
Momofuku Milk Bar                       9.338989  12.867366  19.833335   
The Meatball Shop                       7.895130  11.628012  17.598271   
sweetgreen                             11.383622  16.322607  22.863426   
Blue Ribbon Fried Chicken              12.855661  16.073185  19.225953   
Blue Ribbon Sushi                      11.299781  18.842653  27.054707   
Parm                                    9.338565  14.343995  23.589052   
Whole Foods Market                     23.903119  32.827682  48.211890   
Chipotle Mexican Grill $1.99 Delivery  10.829003  13.664622  16.746878   
Mighty Quinn's BBQ                     13.309285  19.307378  24.331391   
RedFarm Broadway                       11.079151  18.326718  28.108637   
Momofuku Noodle Bar                    20.839799  32.560749  37.942978   
Duane Reade                             9.036986  11.750866  19.846104   
McDonald's                             11.292997  12.563238  15.333491   
TAO                                    15.075144  21.828489  32.933141   

                                              max  
Shake Shack                             72.841402  
Momofuku Milk Bar                       50.063765  
The Meatball Shop                       39.615530  
sweetgreen                              73.839974  
Blue Ribbon Fried Chicken              267.954044  
Blue Ribbon Sushi                       52.855736  
Parm                                    43.337183  
Whole Foods Market                     115.826594  
Chipotle Mexican Grill $1.99 Delivery   36.830231  
Mighty Quinn's BBQ                      36.338317  
RedFarm Broadway                        52.032332  
Momofuku Noodle Bar                     56.785158  
Duane Reade                             29.869439  
McDonald's                              23.783853  
TAO                                     73.808479

* Shake Shack - top Jumpman23 vendor - has median waiting/pickup time of 23 mins and std dev of about 10 mins. Oportunity to streamline process by levaraging Shake Shack's online order form. Jumpmen would not have to stand in line if order placed in advance.
* Whole Foods/ Trader Joe's - Highest median wait times. Both have high std dev. For grocery pickup orders, significant time is used to locate items. Training can be provided to familiarize Jumpmen with store layout, this would cut short pickup times at commonly used supermarkets.

##  Top 15 Place Categories by Order Qty

In [27]:
categoryGroup = df.groupby('place_category')
topCat = categoryGroup.item_quantity.sum().sort_values(ascending = False)[:15]
topCat.iplot(kind='bar', title='Popular Categories')

## Transportation Options Analysis

#### Speed Analysis Using Google Maps API

In [28]:
API_key = 'AIzaSyAmk6GXlkxBUY6Xv-EudwP0-UvTzaiZZG8' #enter Google Maps API key
gmaps = googlemaps.Client(key=API_key)

In [29]:
def distanceFunction(dataset):
    if dataset['vehicle_type'] in ['car','van','scooter','truck','motorcycle']: vType = 'driving'
    elif dataset['vehicle_type'] =='bicycle':vType = 'bicycling'
    else: vType = 'walking'
    origin = (dataset['pickup_lat'],dataset['pickup_lon'])
    destination = (dataset['dropoff_lat'],dataset['dropoff_lon'])
    dist = gmaps.distance_matrix(origin, destination, mode=vType)["rows"][0]["elements"][0]["distance"]["value"]
    return dist*0.000621371 #convert meters to miles

In [32]:
df['distance_in_miles'] = df.apply(distanceFunction, axis = 1)

In [33]:
df['speed_in_mph'] = df['distance_in_miles']/(df['pickup_to_dropoff_mins']/60)

In [34]:
vehicleGroup = df.groupby('vehicle_type')
vehicleGroup.speed_in_mph.describe()

count      mean       std       min       25%       50%  \
vehicle_type                                                             
bicycle       3373.0  6.885427  4.605256  0.040813  4.718094  6.368042   
car            967.0  8.135683  4.713026  0.711902  5.162158  7.104362   
motorcycle      18.0  8.830036  5.333008  3.196263  5.375899  7.446897   
scooter         58.0  7.887925  3.633406  1.108550  5.295696  7.584659   
truck           34.0  9.280280  4.728792  3.522230  5.981221  8.032683   
van             60.0  8.074862  7.209582  1.523387  4.724652  6.480620   
walker         209.0  5.718202  3.200391  0.738920  4.014461  5.235213   

                    75%         max  
vehicle_type                         
bicycle        8.187687  142.602292  
car            9.804812   49.185481  
motorcycle     9.568267   24.165554  
scooter       10.098644   20.650313  
truck         11.096963   26.210541  
van            8.981441   52.545446  
walker         6.890374   26.168965

In [55]:
a=vehicleGroup.speed_in_mph.describe()
a.to_csv('SpeedAnalysis.csv')

* Walker is the slowest means of transportation, Bicycle is the second slowest
* Max speed of 142 mph for bicycle is not possible - seems to be a data integrity issue
* Max speed of 26 mph for walker is not possible - seems to be a data integrity issue

In [35]:
vType = df.vehicle_type.value_counts()
vType

bicycle       3740
car           1050
walker         234
van             69
scooter         64
truck           38
motorcycle      19
Name: vehicle_type, dtype: int64

In [36]:
pieChart = [go.Pie(labels = vType.index, values = vType )]
layout = go.Layout(title= 'Jumpman Vehicle Type')
py.iplot(go.Figure(data=pieChart, layout = layout))

* Despite being a slower option, bicycle is the most popular Jumpman mode of transport. Considering street layout of New York, bicycle presents an efficeint alternative to cars

## Item Count per Order 
    
  

In [37]:
df.item_quantity.value_counts()

1.0     2798
2.0      803
3.0      209
4.0      100
5.0       32
6.0       18
7.0        8
8.0        6
9.0        4
12.0       1
30.0       1
11.0       1
15.0       1
10.0       1
16.0       1
Name: item_quantity, dtype: int64

In [38]:
df.item_quantity.value_counts(normalize = 1)

1.0     0.702309
2.0     0.201556
3.0     0.052460
4.0     0.025100
5.0     0.008032
6.0     0.004518
7.0     0.002008
8.0     0.001506
9.0     0.001004
12.0    0.000251
30.0    0.000251
11.0    0.000251
15.0    0.000251
10.0    0.000251
16.0    0.000251
Name: item_quantity, dtype: float64

* 70% users order 1 item per delivery 
* 20 % users order 2 items per delivery

## Heatmap of Pickup and Drop Locations

In [39]:
pickupMap = gmplot.GoogleMapPlotter(40.744607, -73.990742,13,apikey='AIzaSyB2p8i8-8gIhHjsZOyOU8fmSiHMAZWPmGI')

pickupMap.heatmap(df.pickup_lat, df.pickup_lon,radius = 27,opacity = 0.75)
pickupMap.draw('pickupMap.html')

In [40]:
dropMap = gmplot.GoogleMapPlotter(40.744607, -73.990742,13,apikey='AIzaSyB2p8i8-8gIhHjsZOyOU8fmSiHMAZWPmGI')

dropMap.heatmap(df.dropoff_lat, df.dropoff_lon,radius = 27,opacity = 0.75)
dropMap.draw('dropMap.html')

* Pickup locations are concentrated together whereas drop locations are more spread out
* This could be because Manhattan has certain areas with multiple restaurants concentrated whereas drop locations/ people's homes tend to be more uniformly distributed across the city. In Manhattan, one cannot pinpoint areas with high residential density

## Data Integrity Issues

#### 1) Missing Data

In [57]:
df.isnull().sum()

delivery_id                               0
customer_id                               0
jumpman_id                                0
vehicle_type                              0
pickup_place                              0
place_category                          841
item_name                              1230
item_quantity                          1230
item_category_name                     1230
how_long_it_took_to_order              2635
pickup_lat                                0
pickup_lon                                0
dropoff_lat                               0
dropoff_lon                               0
when_the_delivery_started                 0
when_the_Jumpman_arrived_at_pickup      495
when_the_Jumpman_left_pickup            495
when_the_Jumpman_arrived_at_dropoff       0
delivery_start_date                       0
delivery_start_day                        0
min_to_order                           2635
start_to_pickup_mins                    495
pickup_duration_mins            

* 'How Long it took to order' column has about 50% values missing; Item name and category is missing for 20% of dataset
* Issues with how data entry and management need to be solved to prevent large instances of missing data

#### 2) Incorrect Time Stamp for 'When Delivery Started' and 'When Jumpman arrived at Pickup'


In [42]:
df[df.start_to_pickup_mins<=0]

delivery_id  customer_id  jumpman_id vehicle_type  \
54        1273974       352549      141203      bicycle   
69        1274328       255435       23359      bicycle   
72        1274372        82041      133293      bicycle   
127       1275416        44158       47204      bicycle   
136       1275526       303211       31859      bicycle   
160       1276907       325403      146428          van   
164       1277434        67848      118095      bicycle   
169       1278047        63890      136434      bicycle   
176       1278237       158800      125160      bicycle   
177       1278248        49010      136434      bicycle   
183       1278357       213877       44356      bicycle   
193       1278588       186364      123427      bicycle   
195       1278702        69953      105427      bicycle   
208       1279068        80328       84093      bicycle   
217       1279587       331128      140684      bicycle   
218       1279636        77736      115403      bicycle   
221       1279862       124972       20962       walker   
225       1280134       109405       64664      bicycle   
237       1280752       156461      143427      bicycle   
250       1281246        54365      108647      bicycle   
267       1281585        47440        9078      bicycle   
290       1282090        94936       15291          car   
318       1283820       301032      116326          car   
320       1283828       171192      151073      bicycle   
344       1285013       240982       23359      bicycle   
360       1285292        62730      144899      bicycle   
375       1286240        72924      139444      bicycle   
376       1286310        59765       88804      bicycle   
377       1286394        79715       22502      bicycle   
397       1287603        55334       39202      bicycle   
...           ...          ...         ...          ...   
5006      1482678        62632       61900      bicycle   
5009      1482915       403718        6873          car   
5013      1483276       232035      165830          van   
5017      1483371       102210       67430      bicycle   
5019      1483458       124306      116326          car   
5022      1483591        92768       67430      bicycle   
5039      1485301       329315       66416      bicycle   
5051      1485688       356678      125052      bicycle   
5058      1485928       196101      172700      bicycle   
5064      1486055        66373       84093      bicycle   
5069      1486159        69799      102837          car   
5070      1486160       152675      166619      bicycle   
5072      1486174       366421       69993      bicycle   
5079      1486298       264038      161162      bicycle   
5080      1486342       116293       48430      bicycle   
5091      1486840       143274      170848      bicycle   
5099      1487139       175845      171206          van   
5101      1487179       138586       46764      bicycle   
5112      1487890       151333      181543      bicycle   
5125      1488346       159449      138061          car   
5129      1488426       177250      116881      bicycle   
5132      1488520       135337      125504      bicycle   
5133      1488523        61250      137090      bicycle   
5136      1488577        48131      116881      bicycle   
5141      1488767        52327      145502          car   
5144      1488796        40410      140684      bicycle   
5168      1489274        42706      117223      bicycle   
5203      1490802        70541      161459      bicycle   
5206      1490893       295078      157975      bicycle   
5207      1491029       405547      142394      bicycle   

                                pickup_place  place_category  \
54                             Levain Bakery          Bakery   
69                              Lure Fishbar         Seafood   
72                                      Parm         Italian   
127                                   Chop't           Salad   
136                        

* There are 438 rows (8% of all orders) where time stamp for 'when delivery started' is after 'when jumpman arrived at pickup'


#### 3. Incorrect Time Stamp for 'When Jumpman left pickup' and 'When Jumpman arrived at dropoff' OR Incorrect Coordinates

In [43]:
df[((df.vehicle_type =='bicycle') & (df.speed_in_mph >= 20))]

delivery_id  customer_id  jumpman_id vehicle_type  \
354       1285171       184020       74692      bicycle   
425       1288316       128243       89004      bicycle   
455       1289093       334373       89004      bicycle   
482       1290499       253195       83565      bicycle   
558       1294583       122852      137297      bicycle   
659       1299003        94426       36664      bicycle   
760       1302571       295442      137297      bicycle   
896       1307794       361509      146510      bicycle   
1000      1311664       115231      104533      bicycle   
1522      1330286        53947      148770      bicycle   
1534      1331230       137739       62860      bicycle   
1553      1332223       371590      137090      bicycle   
1593      1333156       371884       83117      bicycle   
1620      1333837        81084      113822      bicycle   
1859      1345301       240014       62860      bicycle   
2040      1353563        64141       31156      bicycle   
2297      1362826        78646      148895      bicycle   
2539      1373207       195714      157975      bicycle   
2608      1375728       150840      109259      bicycle   
2715      1381438        68942       30743      bicycle   
2847      1388047        64618      158962      bicycle   
2864      1388615       167828      158962      bicycle   
3218      1402292       195580       32580      bicycle   
3381      1410394       313639       62860      bicycle   
3693      1423670       370096      100390      bicycle   
3696      1423843        84457       89004      bicycle   
3937      1436805        50616      123844      bicycle   
4094      1444597        91817      171554      bicycle   
4097      1444826       359025       31156      bicycle   
4501      1464027        67844      158962      bicycle   
4520      1464966       259341      158962      bicycle   
4533      1466377       175290      158962      bicycle   
4567      1467797       163738       60761      bicycle   
4712      1473113        41136        9476      bicycle   
4849      1476282        58205      158962      bicycle   
4988      1481866       172624       83117      bicycle   
5085      1486536       145674      123844      bicycle   
5166      1489231       106409      104533      bicycle   

                               pickup_place      place_category  \
354                      Whole Foods Market       Grocery Store   
425                       The Meatball Shop             Italian   
455                      Mighty Quinn's BBQ                 BBQ   
482                              Red Bamboo          Vegetarian   
558                               BLT Prime               Steak   
659                       Momofuku Milk Bar             Dessert   
760                          Luke's Lobster             Seafood   
896                             Shake Shack              Burger   
1000                            Izakaya Ten            Japanese   
1522                      Gaia Italian Cafe             Italian   
1534                          Via Quadronno                 NaN   
1553                      The Meatball Shop             Italian   
1593                               The Palm               Steak   
1620                            Joe's Pizza               Pizza   
1859                             Sarabeth's            American   
2040           Chloe's Soft Serve Fruit Co.             Dessert   
2297                     Mighty Quinn's BBQ                 BBQ   
2539                 Doughnut Plant Chelsea               Donut   
2608                    Momofuku Noodle Bar                 NaN   
2715                                  Petco  Pet Supplies Store   
2847        Saigon Vietnamese Sandwich Deli                 NaN   
2864                            Shake Shack              Burger   
3218                 Postmates Liquor Store                Shop   
3381                      Momofuku Milk Bar             Dessert   
3693                         

* Average bicycle speed of over 20 mph is almost impossible to acheive thus either the time stamps or the coordinates are incorrect

In [44]:
df[(df.vehicle_type =='walker') & (df.speed_in_mph >= 15)]

delivery_id  customer_id  jumpman_id vehicle_type  \
1145      1315957        61583       20962       walker   
3271      1404992       351561       59348       walker   
3950      1437622       262058      140794       walker   
4077      1442582        67316       20962       walker   
4307      1456853       314808       12779       walker   

                pickup_place place_category              item_name  \
1145  Postmates Liquor Store           Shop        Antiyal 'Kuyen'   
3271      Mighty Quinn's BBQ            BBQ             Spare Ribs   
3950             Dos Caminos        Mexican            Traditional   
4077  Postmates Liquor Store           Shop  Grey Goose Vodka (1L)   
4307      The Mercer Kitchen            NaN     Egg White Omelette   

      item_quantity item_category_name  how_long_it_took_to_order  ...  \
1145            2.0       Syrah/Shiraz                        NaT  ...   
3271            1.0              Meats                        NaT  ...   
3950            1.0    Lunch Guacamole 1900-01-01 00:05:20.150709  ...   
4077            2.0              Vodka                        NaT  ...   
4307            1.0   Breakfast (Eggs) 1900-01-01 00:36:59.652869  ...   

      when_the_Jumpman_left_pickup  when_the_Jumpman_arrived_at_dropoff  \
1145    2014-10-07 22:02:55.418596           2014-10-07 22:20:04.458817   
3271    2014-10-19 22:54:13.714259           2014-10-19 23:05:44.707223   
3950    2014-10-24 16:12:08.710530           2014-10-24 16:17:59.291946   
4077    2014-10-24 22:58:41.952169           2014-10-24 23:15:34.713797   
4307    2014-10-26 13:52:12.848720           2014-10-26 13:56:21.853344   

      delivery_start_date  delivery_start_day min_to_order  \
1145          07 Oct, Tue             Tuesday          NaN   
3271          19 Oct, Sun              Sunday          NaN   
3950          24 Oct, Fri              Friday     5.333333   
4077          24 Oct, Fri              Friday          NaN   
4307          26 Oct, Sun              Sunday    36.983333   

     start_to_pickup_mins pickup_duration_mins pickup_to_dropoff_mins  \
1145             7.382937             4.834679              17.150670   
3271            56.778582            16.333788              11.516549   
3950            24.291578            19.811371               5.843024   
4077             5.753910             4.835004              16.879360   
4307            19.330948            64.329624               4.150077   

     distance_in_miles speed_in_mph  
1145          4.520474    15.814451  
3271          4.180584    21.780399  
3950          2.002057    20.558439  
4077          4.255149    15.125509  
4307          1.810054    26.168965  

[5 rows x 26 columns]

* Average walking speed of over 15 mph is almost impossible to acheive thus either the time stamps or the coordinates are incorrect